### ISSUE TITLE
Block content page and internationalization with language prefix

### ISSUE THREAD

**Issue by saintger** [django-oscar/django-oscar#1481]
Hello,

In reference to this post in the mailing-list:
https://groups.google.com/d/msg/django-oscar/ryo6IkzuThY/OHLqUjJgRyUJ

Using Django i18n_patterns to add language prefix to the URL makes things difficult to add block content to a page (for each language, you have to add the corresponding page with the correct language prefix).

It seems that it is related to a bug in Django core in which the django.core.urlresolvers.resolve only resolve URL with the current language prefix (even in other language prefix are correct).

A patch diff being worth a thousand words, here is a naive workaround for Oscar:

```diff
--- a/oscar/apps/promotions/context_processors.py   Fri Aug 29 01:57:26 2014 +0200
+++ b/oscar/apps/promotions/context_processors.py   Sat Sep 06 21:39:43 2014 +0200
@@ -1,6 +1,7 @@
 from itertools import chain

 from oscar.apps.promotions.models import PagePromotion, KeywordPromotion
+from django.utils.translation import get_language


 def promotions(request):
@@ -24,9 +25,14 @@
     """
     Return promotions relevant to this request
     """
+    language = get_language()
+    if language and request.path.startswith('/%s/' % language):
+        page_url = request.path[len('/%s/' % language)-1:]
+    else:
+        page_url = request.path
     promotions = PagePromotion._default_manager.select_related() \
         .prefetch_related('content_object') \
-        .filter(page_url=request.path) \
+        .filter(page_url=page_url) \
         .order_by('display_order')

     if 'q' in request.GET:
diff -r a1248ccfe775 oscar/core/validators.py
--- a/oscar/core/validators.py  Fri Aug 29 01:57:26 2014 +0200
+++ b/oscar/core/validators.py  Sat Sep 06 21:39:43 2014 +0200
@@ -4,7 +4,7 @@
 from oscar.core.loading import get_model
 from django.http import Http404
 from django.utils.translation import ugettext_lazy as _
-
+from django.utils.translation import get_language

 class ExtendedURLValidator(validators.URLValidator):

@@ -29,6 +29,9 @@

     def validate_local_url(self, value):
         value = self.clean_url(value)
+        language = get_language()
+        if language and not value.startswith('/%s/' % language):
+            value = '/%s%s' % (language, value)
         try:
             resolve(value)
         except Http404:
```

**Comment by sasha0** [comment]
After closer look I came to conclusion that we rather need to improve URL field validation for the local URLs, than promotions loading.

If we have promotions for pages `/` and `/de/` (in English and German languages), there are few scenarios how they can be loaded:

- Load using locale-neutral page URL (`/`) will load always _only_ English content block.
- Load using both locale-neutral and locale-specific URLs will load content in _both_ languages, which might be not relevant for the customers.
- Load firstly using locale-specific URL (`/de/`) and if no relevant records, fallback to local-neutral URL (`/`). In this case, it will work as initially intended, but we won't have control where we want to display content block and where not and just display everywhere the same content block.

We already can attach content block to any number of pages, but the URL validator respects only current locale, so that you can attach promotion to `/en-gb/` page on default locale and not `/de/` and visa versa, which is the main issue, I believe.

**Comment by sasha0** [comment]
Fixed by PR #2542
